### DISCUSSION TITLE
EC2 Metadata roleName Request Returned Error

### DISCUSSION THREAD

**Discussion by rookinthenorth** [aws/graph-explorer#77] [Q&A]
**I]. Overview**
1. I am trying to set up AWS Graph Explorer to connect with our AWS Neptune DB. Goal: Visualize and interact with data in AWS Neptune.
2. Running `docker run -p 80:80 -p 443:443 --env HOST={hostname-or-ip-address} graph-explorer` throws an **EC2 Metadata roleName request returned** error.

**II]. Description of Docker Logs**
```
[ec2-user@<EC2 Public IPv4 Address>]$ sudo docker run -p 80:80 -p 443:443 --env HOST=<Private IPv4 Address> graph-explorer
Generating new self-signed SSL cert using <Private IPv4 Address>...
Generating a 2048 bit RSA private key
.................+++
....................+++
writing new private key to 'rootCA.key'
-----
Generating RSA private key, 2048 bit long modulus
..........................................................................+++
...+++
e is 65537 (0x10001)
Signature ok
subject=/C=US/ST=Washington/L=Seattle/O=Graph Explorer/OU=Graph Explorer Dev/CN=Amazon Neptune
Getting CA Private Key
Starting graph explorer...

> graph-explorer@1.1.0 start:proxy-server /graph-explorer
> pnpm -F graph-explorer-proxy-server start

> graph-explorer-proxy-server@1.1.0 start /graph-explorer/packages/graph-explorer-proxy-server
> node node-server.js

No master credentials available Error: EC2 Metadata roleName request returned error
    at IncomingMessage.<anonymous> (/graph-explorer/node_modules/.pnpm/aws-sdk@2.1249.0/node_modules/aws-sdk/lib/util.js:912:34)
    at IncomingMessage.emit (node:events:525:35)
    at IncomingMessage.emit (node:domain:489:12)
    at endReadableNT (node:internal/streams/readable:1358:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  statusCode: 404,
  retryable: false,
  time: 2023-03-11T12:19:01.401Z
}
        Proxy server located at https://localhost
```

 
**III]. To Reproduce**
Details of the infra/systems/tools used:
 - OS: Amazon Linux 2
 - Instance: AWS EC2
 - Date of Error: 11/March/2023

1. Set up Neptune. Enable SG to allow 'All traffic' (as a last ditch effort right now - as the Graph Explorer was not working)
2. Set up Ec2 (t2.medium)
	2.a. Verify Ec2 can connect to and access Neptune: https://github.com/aws/graph-notebook/tree/main/additional-databases/neptune
	2.b. Install docker: https://www.cyberciti.biz/faq/how-to-install-docker-on-amazon-linux-2/
	2.c. Install Git: https://www.how2shout.com/linux/how-to-install-git-on-aws-ec2-amazon-linux-2/
	2.d. Start Docker: https://stackoverflow.com/a/64310814
3. Set up Graph Explorer on EC2: https://github.com/aws/graph-explorer
	3.a. Cloned Graph Explorer onto EC2
	3.b. Built the docker image (docker build -t graph-explorer .)
	3.c. Ran the following command and I get the above error
-- Do note: instead of using the EC2 instance's Private IP Address in Step 3.c., I have tried with the EC2 instance's Public IP Address and even localhost - it fails regardless.
Would request the team's urgent help on this so that I can work with AWS Graph Explorer itself for the current project.

**Comment by triggan** [comment]
A couple of clarifications:  Does your Neptune cluster have IAM enabled?  If so, does your EC2 instance have the proper role attached with permissions to access the Neptune cluster?

**Reply by rookinthenorth** [reply]
Hi @triggan,
Thanks for the quick revert. 
Do you mean if Neptune has IAM DB Auth enabled? If yes - the IAM DB Auth is not enabled for the Cluster at my end 
![image](https://user-images.githubusercontent.com/127603046/225693150-2ebde3bd-a051-43d2-8180-a65d4713ab74.png)

**Reply by rookinthenorth** [reply]
@triggan - any chance you can help point me towards the IAM policy/policies you believe EC2 would need in this case?
Thanks in advance! :)

**Reply by rookinthenorth** [reply]
also - the EC2 does not have any IAM Role associated with it either.

**Reply by angelos-c** [reply]
@rookinthenorth did you manage to solve this? facing the same problem/error
